### PR TITLE
Inference across spatial areas

### LINKED ISSUE #5: Inference across spatial areas
- [x] Define spatial areas which are prone to artefacts in the currently deployed system
- [x] Export tif files for those areas
- [ ] Run inference with the trained model against those areas

### PR DESCRIPTION
**Pull request by gabrieltseng**
This has the potential for conflicts with #13 , so I'll wait for that to get merged in before diving properly into the evaluation code.

- [x] Spatial inference dataset
- [x] Integrate into eval task
- [x] Plot and log results to wandb

Closes #5

### CODE REVIEW

File: presto/utils.py [hunk 1]
```diff
@@ -275,3 +276,16 @@ def plot_for_group(grp_df):
     )
 
     metrics_df.groupby(["model", "metric_type"]).apply(plot_for_group)
+
+
+def plot_spatial(spatial_preds: xr.Dataset, output_path: Path, to_wandb: int = False):
```
Comment (gabrieltseng) [inline presto/utils.py:281]: All plots in [this comment](https://github.com/WorldCereal/presto-worldcereal/pull/14#issuecomment-1881287660) were made using this function.

File: tests/test_eval.py [hunk 2]
```diff
@@ -20,3 +25,28 @@ def test_eval(self):
...
+            data_dir / f"inference_areas/{spatial_data_prefix}.nc", decode_times=False
+        )
+        ground_truth_one_timestep = spatial_data.worldcereal_cropland.values[0, :, :]
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            eval_task = WorldCerealEval(test_data, test_data, Path(tmpdirname))
+            finetuned_model = eval_task._construct_finetuning_model(model)
+            eval_task.spatial_inference(finetuned_model, None)
+            output = xr.open_dataset(Path(tmpdirname) / f"{spatial_data_prefix}_finetuning.nc")
+            # np.flip because of how lat lons are stored vs x y
+            self.assertTrue(
```
Comment (kvantricht) [inline tests/test_eval.py:50]: This tests if "ground truth" is loaded correctly, but we don't test if the spatial inference itself returns the expected results? Is that what we want?

Comment (gabrieltseng) [reply tests/test_eval.py:50]: Testing the inference is a bit more complicated, since we don't know for sure what the expected values are. This lets us test the data loading (and turning the batches back into an `xr.Dataset`) is happening correctly. 

Ideally we would test the inference too, but visually it does seem to make sense and I am not certain how I would test that.

File: presto/eval.py [hunk 3]
```diff
@@ -164,8 +163,61 @@ def evaluate(
...
+                    torch.from_numpy(latlons).float(),
+                    torch.from_numpy(months).long(),
+                    torch.from_numpy(mask).float(),
+                ),
+                batch_size=8192,
+                shuffle=False,
+            )
+            test_preds_np, _ = self._inference_for_dl(dl, finetuned_model, pretrained_model)
+            df = ds.combine_predictions(latlons, test_preds_np, y)
+            if pretrained_model is None:
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply presto/eval.py:194]: `pretrained_model` is None if we finetune Presto.

If we fix the Presto encodings and train a sklearn model on top of them, then `pretrained_model is not None`.

That flow is [here](https://github.com/WorldCereal/presto-worldcereal/blob/main/presto/eval.py#L140).

File: presto/dataset.py [hunk 4]
```diff
@@ -141,13 +147,111 @@ def __getitem__(self, idx):
...
+            elif presto_val == "total_precipitation":
+                # scaling, and AgERA5 is in mm, Presto expects m
+                values = values / (100 * 1000.0)
+            elif presto_val == "temperature_2m":
+                # remove scaling
+                values = values / 100
+
+            eo_data[:, :, BANDS.index(presto_val)] = values
+
+            mask[:, :, IDX_TO_BAND_GROUPS[presto_val]] = np.clip(
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply presto/dataset.py:218]: In these lines we are creating the binary mask that tells Presto which values to ignore (`~idx_valid`).

The `.clip` is just to make sure the mask never has values greater than `1`.

File: presto/dataset.py [hunk 5]
```diff
@@ -141,13 +147,111 @@ def __getitem__(self, idx):
...
+        cls, filepath: Path
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        ds = cast(xr.Dataset, rioxarray.open_rasterio(filepath, decode_times=False))
+        epsg_coords = ds.rio.crs.to_epsg()
+
+        num_instances = len(ds.x) * len(ds.y)
+        num_timesteps = len(ds.t)
+        eo_data = np.zeros((num_instances, num_timesteps, len(BANDS)))
+        mask = np.zeros((num_instances, num_timesteps, len(BANDS_GROUPS_IDX)))
+        # for now, B8A is missing
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply presto/dataset.py:198]: Is it an easy addition to the data? It might be an interesting experiment to see if the model performs much better with it.

However for now I am happy to keep it out so that we still have an apples-to-apples comparison with the CatBoost model.

File: tests/test_eval.py (hunk 2 above)
Comment (kvantricht) [reply tests/test_eval.py:50]: Agreed, but I was just thinking if it makes sense to not only compare ground truth, but also the (flattened) predictions of spatial inference. Just to make sure the pipeline works as expected. We could do that just using the pretrained model, because the exact values of the predictions don't matter, as long as we verify that that they always return the same result?

File: presto/dataset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kvantricht) [reply presto/dataset.py:218]: In which situation would such values be greater than `1` and would `clip` do exactly this without warning that it doesn't seem a normal situation that these original values were outside the expected range?

[the remark this answers is not in the document]
Comment (gabrieltseng) [reply presto/dataset.py:218]: Yes this is probably excessively defensive - I copied this over from the [original Eval tasks](https://github.com/nasaharvest/presto/blob/main/presto/eval/eurosat_eval.py#L179). 

Originally, this was necessary because the dataloader would return a mask *and* the user can provide their own mask. So if the user wants to mask out a value which the dataloader has already masked, then the `clip` would be necessary. 

However, it can't happen in this case so its safe to remove.

**Comment by gabrieltseng** [comment]
Added thresholded predictions to the plots for easier comparison to the worldcereal values (this is the Sudan tile):

File: presto/dataset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply presto/dataset.py]: Removed
